src/assets/wise5/components/component-student.component.ts

```diff
@@ -9,18 +17,463 @@ export abstract class ComponentStudent {
   @Input()
   componentContent: any;
 
+  @Input()
+  componentState: any;
+
+  @Input()
+  mode: string;
+
+  @Input()
+  workgroupId: number;
+
   componentId: string;
+  componentType: string;
+  prompt: string;
+  isPromptVisible: boolean = true;
+  isSaveButtonVisible: boolean = false;
+  isSubmitButtonVisible: boolean = false;
+  isSaveOrSubmitButtonVisible: boolean = false;
+  isSubmitButtonDisabled: boolean = false;
+  isSubmitDirty: boolean = false;
+  isSubmit: boolean = false;
+  isDirty: boolean = false;
+  isDisabled: boolean = false;
+  submitCounter: number = 0;
+  latestAnnotations: any;
+  saveMessage = {
+    text: '',
+    time: ''
+  };
+  requestComponentStateSubscription: Subscription;
+  annotationSavedToServerSubscription: Subscription;
+  nodeSubmitClickedSubscription: Subscription;
+  studentWorkSavedToServerSubscription: Subscription;
 
-  constructor(protected NodeService: NodeService) {}
+  constructor(
+    protected AnnotationService: AnnotationService,
+    protected ComponentService: ComponentService,
+    protected ConfigService: ConfigService,
+    protected NodeService: NodeService,
+    protected StudentDataService: StudentDataService,
+    protected UtilService: UtilService
+  ) {}
 
-  ngOnInit() {
+  ngOnInit(): void {
     this.componentId = this.componentContent.id;
+    this.componentType = this.componentContent.type;
+    this.prompt = this.componentContent.prompt;
+    if (!this.isAuthoringComponentPreviewMode()) {
+      this.latestAnnotations = this.AnnotationService.getLatestComponentAnnotations(
+        this.nodeId,
+        this.componentId,
+        this.workgroupId
+      );
+    }
+    this.registerListeners();
+    this.subscribeToRequestComponentState();
+  }
+
+  ngOnDestroy(): void {
+    this.annotationSavedToServerSubscription.unsubscribe();
+    this.nodeSubmitClickedSubscription.unsubscribe();
+    this.requestComponentStateSubscription.unsubscribe();
+    this.studentWorkSavedToServerSubscription.unsubscribe();
```

Comment:
```suggestion
    this.subs.unsubscribe();
```
Use Subscription.add and Subscription.unsubscribe to simplify code.